### DISCUSSION TITLE
Check if the query is already executed and resolved

### DISCUSSION THREAD

**Discussion by samislam** [Automattic/mongoose#12177] [Q&A]
I have the following code:
```js
const query = UserModel.find({})
const resolvedValue = await query
```
I want to check if the `query` was executed, how do I check that? is there something like the following I can use in mongoose?
```js
console.log(query.resolved) // false or true
// or
console.log(query.awaited) // false or true
// or
console.log(query.executed) // false or true
```
I know that I can simply see if the variable `resolvedValue` is not undefined, but I want to check if there's such an option in mongoose, which would be pretty handy if it exists.

**Comment by vkarpov15** [comment]
There isn't a good publicly facing way to do this, other than to `await query` and see if that throws an error.

Why are you looking to check whether a given query has already been executed?

**Reply by samislam** [reply]
Because I have written a package called setDoc, and the way I was planning to make it is:
it requires the user to write a function which accepts two types as the first argument, and a second argument which is the options argument:
- function.
- asyncfunction.
The consumer of the function will use it as follows;
```js
setDocMw( ()=> Model.find(), { options } )
// or for example
setDocMw( async (req)=> (await getModelById(req.body.recipient)).findById(req.params.id), {
pre: (query)=> query.sort() // used to chain methods on the query before it gets awaited
} )
```

That's how it looks like from the perspective of the consumer of the package, 
but from inside. when a user provides an async function as the first argument, I must await that async function to receive the **query** (the returned value from the async function). 
However, awaiting the async function awaits the query, (_why? so strange..._)

I should await the query, because I need the query to chain some methods on it, 
for instance, the `pre` option is used for that, and in order for me (the developer) to pass the argument `query` as the first parameter to the `pre` option method, I need to have the `query`, not the resolved documents.

and the only way I found to achieve this pattern is by having an option like `query.isResolved?` or `query.isAwaited` or `query.isExecuted`.

I have written the package however, I wanted at first to have that pre/post options, but it did not work, so I have chosen another pattern for that.

The package is on [npm](https://www.npmjs.com/package/setdoc) as well as [github](https://github.com/samislam/setdoc#readme)

**Reply by vkarpov15** [reply]
The workaround is to `return { query }` instead of `return query` from an async function. Because Mongoose queries have a `then()` function, returning a query from an async function automatically executes the query. Does that work for you?

**Reply by samislam** [reply]
yes, that's right.

**Reply by samislam** [reply]
I'll make a request for a new feature for this, however, ok?

**Reply by vkarpov15** [reply]
I'm not sure how the new feature would work. The only option would be to remove the `then()` function from Mongoose queries, which would drop support for promise chaining and async/await. And we won't do that.
